gen/service.go

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package gen
+
+import (
+	"bytes"
+	"fmt"
+	"go/token"
+	"strings"
+
+	"github.com/thriftrw/thriftrw-go/compile"
+)
+
+// Service generates code for the given service.
+//
+// Returns a map from file name to contents for that file. The file names are
+// relative to the package directory for the service.
+func Service(g Generator, s *compile.ServiceSpec) (map[string]*bytes.Buffer, error) {
+	files := make(map[string]*bytes.Buffer)
+
+	// TODO inherited service functions
+
+	for _, functionName := range sortStringKeys(s.Functions) {
+		err := ServiceFunction(g, s, s.Functions[functionName])
+		if err != nil {
+			return nil, fmt.Errorf(
+				"could not generate types for %s.%s: %v",
+				s.Name, functionName, err)
+		}
+
+		buff := new(bytes.Buffer)
+		if err := g.Write(buff, token.NewFileSet()); err != nil {
+			return nil, fmt.Errorf("could not write %s.%s: %v", s.Name, functionName, err)
+		}
+
+		// TODO check for conflicts
+		files[strings.ToLower(functionName)+".go"] = buff
+	}
+
+	return files, nil
+}
+
+// ServiceFunction generates code for the given function of the given service.
+func ServiceFunction(g Generator, s *compile.ServiceSpec, f *compile.FunctionSpec) error {
+	if f.OneWay {
+		return fmt.Errorf(
+			"oneway functions are not yet supported: %s.%s is oneway",
+			s.Name, f.Name)
+	}
+
+	argsGen := fieldGroupGenerator{
+		Name:   goCase(f.Name) + "Args",
```

Comment:
As discussed, appended `Helper` to the variable name. We can iterate on this if we decide this isn't good enough for the plugins.